core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java

```diff
@@ -13,22 +13,55 @@
  */
 package io.trino.spi.classloader;
 
-import java.io.Closeable;
+import java.util.function.Supplier;
 
 public class ThreadContextClassLoader
-        implements Closeable
 {
-    private final ClassLoader originalThreadContextClassLoader;
+    private ThreadContextClassLoader() {}
 
-    public ThreadContextClassLoader(ClassLoader newThreadContextClassLoader)
+    public static <T> T withClassLoader(ClassLoader newThreadContextClassLoader, Supplier<T> supplier)
     {
-        this.originalThreadContextClassLoader = Thread.currentThread().getContextClassLoader();
+        final ClassLoader originalThreadContextClassLoader = Thread.currentThread().getContextClassLoader();
         Thread.currentThread().setContextClassLoader(newThreadContextClassLoader);
+        try {
+            return supplier.get();
+        }
+        finally {
+            Thread.currentThread().setContextClassLoader(originalThreadContextClassLoader);
+        }
     }
 
-    @Override
-    public void close()
+    public static void withClassLoader(ClassLoader newThreadContextClassLoader, Runnable runnable)
     {
-        Thread.currentThread().setContextClassLoader(originalThreadContextClassLoader);
+        withClassLoader(newThreadContextClassLoader, () -> {
+            runnable.run();
+            return null;
+        });
+    }
+
+    public static <T, E extends Exception> T withClassLoader(ClassLoader newThreadContextClassLoader, ThrowingSupplier<T, E> supplier)
+    {
+        return withClassLoader(newThreadContextClassLoader, (Supplier<T>) supplier);
+    }
+
+    @FunctionalInterface
+    public interface ThrowingSupplier<T, E extends Exception>
+            extends Supplier<T>
+    {
+        @Override
+        default T get()
+        {
+            try {
+                return supplyOrThrow();
+            }
+            catch (RuntimeException e) {
+                throw e;
+            }
+            catch (Exception e) {
+                throw new RuntimeException(e);
```

Comment:
We should have ThrowingSupplier that just throws, so that exceptions are propagated as-is,

Once you do that `E extends Exception` -> `E extends Throwable`.

Reply:
It leads to another round of "Checked vs. Unchecked" fight:

<img width="580" alt="Screenshot 2021-07-22 at 11 29 03" src="https://user-images.githubusercontent.com/8093171/126617911-908ce2a1-3d2c-44cf-a712-690d3320d5b0.png">

Reply:
`ThrowingSupplier` must not be a `Supplier`  (as it is on your screenshot)

Reply:
makes sense, PTAL: https://github.com/trinodb/trino/pull/8507/commits/e60dfc9d2b3e1fc80f1d3e4520670f890c8acf1d